alf/experience_replayers/replay_buffer.py

```diff
@@ -254,7 +362,187 @@ def _prioritized_sample(self, batch_size, batch_length):
         r = (r + torch.arange(batch_size, dtype=torch.float32)) / batch_size
         r = r * self._sum_tree.summary()
         indices = self._sum_tree.find_sum_bound(r)
-        return self._index_to_env_id_pos(indices)
+        env_ids, idx = self._index_to_env_id_idx(indices)
+        return env_ids, idx
+
+    def _pad(self, x, env_ids):
+        """Make ``x`` (any index) absolute positions in the RingBuffer.
+
+        This is the reverse of ``circular()``, and is useful when trying to
+        compute the distance from one index to the other.  This is done by
+        adding multiples of _max_length back to the index.
+
+        NOTE, this operation depends on the _current_pos of the RingBuffer,
+        and can generate different results if new data are added to the buffer.
+        """
+        multiples = (
+            self._current_pos[env_ids] / self._max_length) * self._max_length
+        idx = self.circular(x)
+        return torch.where(idx < self.circular(self._current_pos[env_ids]),
+                           idx + multiples, idx + multiples - self._max_length)
+
+    def store_episode_end_pos(self, non_first, pos, env_ids):
+        """Update _indexed_pos and _headless_indexed_pos for episode end pos.
+
+        Args:
+            non_first_idx (tensor): index of the added batch of exp, which are
+                not FIRST steps.  We need to update last step pos for all
+                these env_ids[non_first_idx].
+            pos (tensor): position of the stored batch.
+            env_ids (tensor): env_ids of the stored batch.
+        """
+        _env_ids = env_ids[non_first]
+        _pos = pos[non_first]
+        # Because this is a non-first step, the previous step's first step
+        # is the same as this stored step's first step.  Look it up.
+        prev_pos = _pos - 1
+        prev_idx = self.circular(prev_pos)
+        prev_first = self._indexed_pos[(_env_ids, prev_idx)]
+        prev_first_idx = self.circular(prev_first)
+        # Record pos of ``FIRST`` step into the current _indexed_pos
+        self._indexed_pos[(_env_ids, self.circular(_pos))] = prev_first
+
+        # Store episode end into the ``FIRST`` step of the episode.
+        has_head_cond = prev_first > _pos - self._max_length
+        has_head, = torch.where(has_head_cond)
+        self._indexed_pos[(_env_ids[has_head],
+                           prev_first_idx[has_head])] = _pos[has_head]
+        # For a headless episode whose ``FIRST`` step was overwritten by new
+        # data, the current step has to belong to the same episode as all the
+        # other steps in the buffer, i.e. episode is longer than max_length of
+        # the buffer.  This means prev_first <= _pos - max_length.
+        headless, = torch.where(torch.logical_not(has_head_cond))
+        self._headless_indexed_pos[_env_ids[headless]] = _pos[headless]
+
+    def get_episode_end_pos(self, idx, env_ids):
```

Comment:
_get_episode_end_pose to make it private function. Otherwise, need to document idx and env_ids. For a public function, it might be better to use position instead of idx as argument.

Reply:
make it private